src/driver.jl

```diff
@@ -317,6 +316,18 @@ const __llvm_initialized = Ref(false)
     end
 
     @timeit_debug to "IR post-processing" begin
+        # mark the kernel entry-point functions (optimization may need it)
+        if job.source.kernel
+            # XXX: we want to save the actual function here, but due to our passes rewriting
+            #      functions, and the inability to RAUW values with a different type, that
+            #      metadata gets lost. So instead we save the function name. See also:
+            #      https://discourse.llvm.org/t/replacing-module-metadata-uses-of-function/62431
+            push!(metadata(ir)["julia.kernel"], MDNode([MDString(LLVM.name(entry); ctx)]; ctx))
```

Comment:
How about we do it the other way around and add metadata/attribute to the kernel function

Reply:
See https://github.com/maleadt/LLVM.jl/pull/304